src/sqlfluff/core/linter/linter.py

```diff
@@ -489,117 +493,152 @@ def lint_fix_parsed(
         if not config.get("disable_noqa"):
             rule_codes = [r.code for r in rule_set]
             ignore_buff, ivs = cls.extract_ignore_mask_tree(tree, rule_codes)
-            all_linting_errors += ivs
+            initial_linting_errors += ivs
         else:
             ignore_buff = []
 
         save_tree = tree
-        for loop in range(loop_limit):
-            linter_logger.info(f"Linter loop {loop+1}/{loop_limit}")
-            changed = False
-
-            progress_bar_crawler = tqdm(
-                rule_set,
-                desc="lint by rules",
-                leave=False,
-                disable=progress_bar_configuration.disable_progress_bar,
-            )
-
-            for crawler in progress_bar_crawler:
-                progress_bar_crawler.set_description(f"rule {crawler.code}")
-
-                # fixes should be a dict {} with keys edit, delete, create
-                # delete is just a list of segments to delete
-                # edit and create are list of tuples. The first element is the
-                # "anchor", the segment to look for either to edit or to insert BEFORE.
-                # The second is the element to insert or create.
-                linting_errors, _, fixes, _ = crawler.crawl(
-                    tree,
-                    ignore_mask=ignore_buff,
-                    dialect=config.get("dialect_obj"),
-                    fname=fname,
-                    fix=fix,
-                    templated_file=templated_file,
+        # There are two phases of rule running. The main loop is for most rules.
+        # These rules are assumed to interact and cause a cascade of fixes
+        # requiring multiple passes. The post loop is for post-processing rules,
+        # not expected to trigger any downstream rules, e.g. capitalization fixes.
+        phases = ["main"]
+        if fix:
+            phases.append("post")
+        for phase in phases:
+            if len(phases) > 1:
+                rules_this_phase = [
+                    rule for rule in rule_set if rule.lint_phase == phase
+                ]
+            else:
+                rules_this_phase = rule_set
+            for loop in range(loop_limit if phase == "main" else 2):
+
+                def is_first_linter_pass():
+                    return phase == phases[0] and loop == 0
```

Comment:
Why defining this function inline, It cant be used anywhere else because of scoping issues right?

Reply:
I made it inline so it'd have access to the variables it needs. Making a separate function with 3 parameters that's only used from one place would've made things harder to understand, I think.

There are a few different ways to do this. The main goal was to make it pretty clear which bits of code inside the loop only run the very first time. (Having two phases was making the old logic -- which was already a little messy -- too difficult to follow.)